test/reduce.js

```diff
@@ -0,0 +1,392 @@
+                } catch (ex) {
+                    // AST is not well formed.
+                    // no harm done - just log the error, ignore latest change and continue iterating.
+                    console.error("*** Error generating code from AST.");
+                    console.error(ex.stack);
+                    console.error("*** Discarding permutation and continuing.");
+                }
+                if (code) {
+                    var differs = producesDifferentResultWhenMinified(code, minify_options, timeout);
+                    if (differs) {
+                        if (differs.timed_out) {
+                            // can't trust the validity of `code_ast` and `code` when timed out.
+                            // no harm done - just ignore latest change and continue iterating.
+                            if (timeout < 5000) timeout += 100;
+                        } else {
+                            // latest permutation is valid, so use it as the basis of new changes
+                            testcase_ast = code_ast;
+                            testcase = code;
+                            var testcase_unminified_result = differs.unminified_result;
+                            var testcase_minified_result = differs.minified_result;
+                        }
+                    }
+                }
+            }
+            if (c == 0) break;
+            if (verbose) {
+                console.error("// reduce test pass " + pass + ": " + testcase.length + " bytes");
+            }
+        }
+        reduced = true;
+        testcase += "\n// output: " + testcase_unminified_result
+            + "\n// minify: " + testcase_minified_result
+            + "\n// options: " + minify_options_json;
+    } else {
+        // same stdout result produced when minified
+        testcase = "// Can't reproduce test failure with minify options provided:"
+            + "\n// " + minify_options_json;
+    }
+    var result = U.minify(testcase, {
+        compress: false,
+        mangle: false,
+        output: {
+            beautify: true,
+            braces: true,
+            comments: true,
+        }
+    });
+    result.reduced = reduced;
+    return result;
+};
+
+function producesDifferentResultWhenMinified(code, minify_options, timeout) {
+    var toplevel = undefined;
+    var unminified_result = run_code(code, toplevel, timeout);
+    if (/timed out/i.test(unminified_result)) return false;
+    if (/^\s*$|Error/.test(unminified_result)) return false;
+
+    var minified_result = run_code(U.minify(code, minify_options).code, toplevel, timeout);
+    if (/timed out/i.test(minified_result)) return { timed_out: true };
+    if (/^\s*$/.test(minified_result)) return false;
+
+    return unminified_result !== minified_result ? {
+        unminified_result: unminified_result,
+        minified_result: minified_result,
+    } : false;
+}
+
+Error.stackTraceLimit = Infinity;
```

Comment:
Just curious &minus; is this meant to be left in here?

Reply:
Absolutely. It's very useful. The reduce_test error is often 200 or more frames down. This is a debugging utility, after all.